### ISSUE TITLE
<int>[].runtimeType != const <int>[].runtimeType (even though they print identically)

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#34787]
```dart
void main() {
  var a = new List<int>();
  var b = <int>[];
  var c = const <int>[];
  print(a.runtimeType); // "List<int>"
  print(b.runtimeType); // "List<int>"
  print(c.runtimeType); // "List<int>"
  print(a.runtimeType.toString() == b.runtimeType.toString()); // true
  print(b.runtimeType.toString() == c.runtimeType.toString()); // true
  print(a.runtimeType == b.runtimeType); // true
  print(b.runtimeType == c.runtimeType); // false (!)
}
```

**Comment by lrhn** [comment]
Yes, constant lists do not have the same run-time type as non-constant lists. That's pretty much a given since they behave differently. Their `runtimeType` getter have not been overridden (AFAIK), and they return different `Type` objects representing different types.

The `toString` of a `Type` object is not specified anywhere. It's easy for two different types to have the same `toString` (for example, the library name is not included). In this case, I guess the VM chooses to print the `toString` of the most specific *public* supertype  to avoid showing the name of their internal implementation class, which users can't use anyway (that would be `_GrowableList<int>` and `_ImmutableList<int>`).

So, I'm not sure what the issue here is. Do you want the `toString` return values to be different? If so, what should they be?

I'd just say: don't use the `toString` of a `Type` object for anything, except perhaps printing it when debugging. There are no promises what it will do, especially for internal platform types.

I'd probably also say: Don't use `runtimeType` for anything serious at all, it's overridable and too fragile to be really useful.

**Comment by Hixie** [comment]
We use runtimeType as the core underpinning of Flutter. :-)

This bug is that the current situation is confusing. The output from printing the runtimeType leads one to assume that the types are the same, but they aren't.

**Comment by lrhn** [comment]
And now I'm sure you know how I feel about that choice :)

This brings be back to my question: What should the toString return instead?
It could be `_GrowableList`, `_FixedLenghtList` and `_ImmutableList`, but that's is platform dependent behavior (dart2js prints `JSArray` in every case because the underlying implementation really is the same, it's a JS Array with extra flags on of fixed-length/unmodifiable lists).

The VM isn't completely consistent here, it prints `_InternalLinekdHashMap` and `_ImmutableMap` for mutable and const maps respectively. The list classes have been special cased in the [`GenerateUserVisibleName`](https://github.com/dart-lang/sdk/blob/master/runtime/vm/object.cc#L4014) function since forever, along with some other classes. I guess the map classes should be too.

**Comment by eernstg** [comment]
Why wouldn't it be acceptable for these platform types to have a `runtimeType` which reports the "official" type of the object, keeping the actual implementation type away from the reported `Type` entirely, rather than making that `Type` print differently?:

```dart
bool _pseudoRandom = true;

class PublicInterface<X> {
  factory PublicInterface() {
    _pseudoRandom = !_pseudoRandom;
    return _pseudoRandom
        ? _PrivateImplementation1()
        : _PrivateImplementation2();
  }
}

Type typeOf<X>() => X;

class _PrivateImplementation1<X> implements PublicInterface<X> {
  Type get runtimeType => typeOf<PublicInterface<X>>();
}

class _PrivateImplementation2<X> implements PublicInterface<X> {
  Type get runtimeType => typeOf<PublicInterface<X>>();
}

main() {
  print(
      PublicInterface<int>().runtimeType == PublicInterface<int>().runtimeType);
}
```

**Comment by lrhn** [comment]
It's definitely allowed for a class to override `runtimeType`, and we already do for `_Smi` and `_Mint`.
However, since users are using `runtimeType` to distinguish different classes, you should be careful if you return the same type from obviously different classes. Returning the same type from `_GrowableList` and `ImmutableList` seems weird (and yet, dart2js does so).

But, to get back to an earlier point: We do not promise that different types have different `toString` results. Since the string contains just the simple name of the type, two types with the same name, but from different libraries, would have the same `toString`. You can just consider `_GrowableList`, `_FixedLengthList` and `_ImmutableList` to all be called `List` and be declared in separate libraries.

Since the `toString` of list types have had the same behavior for at least six years (that's how far back in the git history I saw it before I stopped caring), I don't expect us to change it now.

**Comment by Hixie** [comment]
The question here isn't about what we have a right to do, but about what is the right thing to do.

If anything that's been buggy or confusing for six years can't be changed, then we should all go home. :-)

**Comment by lrhn** [comment]
Then I guess I am asking which `toString` results you would prefer to what we have in the future?
We can return anything we want, it's going through C++ code anyway.

Is it only `List` where this is an issue?

Does it matter that the types are not the same between dart2js and the VM? (They are all `List` on dart2js because it is basically the same "class", with flags that coverns its behavior to make it fixed-length or unmodifiable as needed).

We do not promise that different types have different `toString` return values. Dart2js seems to add numbers after types that have the same name, probably because they map all types to a flat JS namespace. The VM does not. That means you cannot trust the name of a type to be consistent between programs.

For example, the `Symbol` implementation class and the `Symbol` interface class are two different classes. If you test with `print((#x).runtimeType == Symbol);` it prints false on both VM and dart2js. Both classes are named `Symbol` and their `Type` objects should return the same string from `toString` (`Symbol`).

Should we change that?

 In dart2js, one of them actually returns `Symbol0`, but you can't control which one. Using the `toString` of a `Type` object is even more fragile than using the `Type` object itself.

**Comment by rakudrama** [comment]
dart2js has even more interesting names when minified.
You might get e.g. `"minified:pK<minified:t>"` instead of `"JSArray<int>"`.
The choice of minified names can change each time the source code changes.
We have no intention of making the unminified names available in the running program as the cost in download size is too much.
